src/lib/operators/insert.cpp

```diff
@@ -8,94 +8,80 @@
+namespace {
 
-// We need these classes to perform the dynamic cast into a templated ValueSegment
-class AbstractTypedSegmentProcessor : public Noncopyable {
- public:
-  AbstractTypedSegmentProcessor() = default;
-  AbstractTypedSegmentProcessor(const AbstractTypedSegmentProcessor&) = delete;
-  AbstractTypedSegmentProcessor& operator=(const AbstractTypedSegmentProcessor&) = delete;
-  AbstractTypedSegmentProcessor(AbstractTypedSegmentProcessor&&) = default;
-  AbstractTypedSegmentProcessor& operator=(AbstractTypedSegmentProcessor&&) = default;
-  virtual ~AbstractTypedSegmentProcessor() = default;
-  virtual void resize_vector(std::shared_ptr<BaseSegment> segment, size_t new_size) = 0;
-  virtual void copy_data(std::shared_ptr<const BaseSegment> source, ChunkOffset source_start_index,
-                         std::shared_ptr<BaseSegment> target, ChunkOffset target_start_index, ChunkOffset length) = 0;
-};
+using namespace opossum;  // NOLINT
 
 template <typename T>
-class TypedSegmentProcessor : public AbstractTypedSegmentProcessor {
- public:
-  void resize_vector(std::shared_ptr<BaseSegment> segment, size_t new_size) override {
-    auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(segment);
-    DebugAssert(value_segment, "Cannot insert into non-ValueColumns");
-    auto& values = value_segment->values();
+void resize_value_segment(std::shared_ptr<BaseSegment> segment, size_t new_size) {
+  const auto value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(segment);
+  Assert(value_segment, "Cannot insert into non-ValueColumns");
 
-    values.resize(new_size);
+  value_segment->values().resize(new_size);
 
-    if (value_segment->is_nullable()) {
-      value_segment->null_values().resize(new_size);
-    }
+  if (value_segment->is_nullable()) {
+    value_segment->null_values().resize(new_size);
   }
+}
 
-  // this copies
-  void copy_data(std::shared_ptr<const BaseSegment> source, ChunkOffset source_start_index,
-                 std::shared_ptr<BaseSegment> target, ChunkOffset target_start_index, ChunkOffset length) override {
-    auto casted_target = std::dynamic_pointer_cast<ValueSegment<T>>(target);
-    DebugAssert(casted_target, "Cannot insert into non-ValueColumns");
-    auto& values = casted_target->values();
-
-    auto target_is_nullable = casted_target->is_nullable();
-
-    if (auto casted_source = std::dynamic_pointer_cast<const ValueSegment<T>>(source)) {
-      std::copy_n(casted_source->values().begin() + source_start_index, length, values.begin() + target_start_index);
+template <typename T>
+void copy(const std::shared_ptr<const BaseSegment>& source_base_segment, ChunkOffset source_begin_offset,
+          const std::shared_ptr<BaseSegment>& target_base_segment, ChunkOffset target_begin_offset,
+          ChunkOffset length) {
+  const auto target_value_segment = std::dynamic_pointer_cast<ValueSegment<T>>(target_base_segment);
+  Assert(target_value_segment, "Cannot insert into non-ValueSegments");
```

Comment:
DebugAssert that the vector has enough space?

Reply:
done